addon/-private/system/model/internal-model.js

```diff
@@ -235,8 +241,68 @@ InternalModel.prototype = {
     return this.record;
   },
 
-  unloadRecord() {
+  directlyRelatedInternalModels() {
+    var array = [];
+    this.type.eachRelationship((key, relationship) => {
+      if (this._relationships.has(key)) {
+        var related = this._relationships.get(key).members.toArray();
+        array.push(...related);
+      }
+    });
+    return array;
+  },
+
+  allRelatedInternalModels() {
+    let array = [];
+    let queue = [];
+    let seen = [];
+    queue.push(this);
+    seen.push(this);
+    while (queue.length > 0) {
+      let node = queue.shift();
+      array.push(node);
+      let related = node.directlyRelatedInternalModels();
+      related.forEach(function(internalModel) {
+        if (seen.indexOf(internalModel) === -1) {
+          queue.push(internalModel);
+          seen.push(internalModel);
+        }
+      });
+    }
+    return array;
+  },
+
+  unloadRecord: function() {
     this.send('unloadRecord');
+    this.dematerializeRecord();
+    var relatedInternalModels = this.allRelatedInternalModels();
+    var allUnloaded = true;
+    var record;
+    for (var i=0; i < relatedInternalModels.length; i++) {
+      record = relatedInternalModels[i].record;
+      if (record && !(record.get('isDestroyed') || record.get('isDestroying'))) {
+        allUnloaded = false;
+        break;
+      }
+    }
+    if (allUnloaded) {
+      relatedInternalModels.forEach((internalModel) => {
+        if (!internalModel.isDestroyed) {
+          internalModel.destroy();
+        }
+      });
+    }
+  },
+
+  destroy: function() {
+    this._isDestroyed = true;
```

Comment:
we should likely mimic embers flags:

* `isDestroying = true` during distruction
* `isDestroyed = true` once destroyed